### PR TITLE
Remote execution of torch hooks

### PR DESCRIPTION
**Pull request by LaRiffle**
## Description
This provides a mechanism to set hook functions (during pytorch backprop) on remote tensors and models.

This allows to register torch hooks on remote tensors. Such operation
        is tricky because you can't really send the hook function to a remote
        party, as python functions are not serializable within PySyft. So you
        need to keep it attached to the PointerTensor.
        On the other hand, the PointerTensor cannot watch for gradient update
        or be triggered natively by torch when backpropagation happens. That's
        why we actually remotely set a hook that we call a callback hook whose
        function is only to call back the pointer during the backpropagation
        to effectively run the hook function.
        So the workflow is: the remote hook is triggered by pytorch, a message
        is sent back to the pointer owner which has the hook function, then
        the hook function is run remotely on the remote gradient, and a termi-
        nation message is returned to the gradient owner.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: syft/frameworks/torch/fl/dataset.py [hunk 1]
```diff
@@ -257,12 +257,22 @@ def get_protobuf_schema():
...
     part and sent to each workers
+
+    Args:
+        dataset (Dataset): the dataset to federate across workers
+        *workers (BaseWorker): the workers receiving parts of the dataset
+        **kwargs (dict): any kwargs to use when sending parts of the dataset
+            to the workers
     """
+    # Compat with the old signature without the (*)workers
+    if isinstance(workers[0], tuple):
```
Comment (gmuraru) [inline syft/frameworks/torch/fl/dataset.py:273]: Q: Can this be a list?

File: syft/frameworks/torch/tensors/interpreters/native.py [hunk 2]
```diff
@@ -89,6 +89,50 @@ def trigger_origin_backward(grad):
...
+        """
+        Register a Torch hook (that is triggered when the `self` tensor has a gradient
+        update) which will send back to a location a specific message whose arguments
+        are updated with a reference to `self`.
+
+        Args:
+            message (Message): the message to send back
+            location (BaseWorker): the worker to which the message should be sent
+        """
+        location = self.owner.get_worker(location)
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/native.py:102]: Q: Could you add tests for those?

File: syft/frameworks/torch/fl/dataset.py [hunk 3]
```diff
@@ -315,6 +325,8 @@ def workers(self):
 
         return list(self.datasets.keys())
 
+    locations = workers
```
Comment (gmuraru) [inline syft/frameworks/torch/fl/dataset.py:328]: Q: Is this used?

File: syft/frameworks/torch/fl/dataset.py (hunk 1 above)
Comment (LaRiffle) [reply syft/frameworks/torch/fl/dataset.py:273]: Yup! Will change

File: syft/frameworks/torch/fl/dataset.py (hunk 3 above)
Comment (LaRiffle) [reply syft/frameworks/torch/fl/dataset.py:328]: It's an alias which is more intuitive, we always have .location(s), except for this one

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 2 above)
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/native.py:102]: Yeah... :)

Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/native.py:102]: @gmuraru done :)

**Review by gmuraru — APPROVED** [review]
